algebird-core/src/main/scala/com/twitter/algebird/Window.scala

```diff
@@ -0,0 +1,118 @@
+/*
+Copyright 2018 Stripe
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package com.twitter.algebird
+
+import scala.collection.immutable.Queue
+
+/**
+ *
+ * Convenience case class defined with a monoid for aggregating elements over
+ * a finite window.
+ *
+ * @param total wrapped instance of `T`
+ * @param q (optional) queue to override default q containing input "total"
+ */
+
+case class Window[T](total: T, q: Queue[T] = Queue()) {
+  def items = if (q.size > 0) q else Queue(total)
+  def size = items.size
+}
+
+/**
+ * Provides a natural monoid for combining windows truncated to some window size.
+ *
+ * @param windowSize
+ */
+
+case class WindowMonoid[T](
+    windowSize: Int
+)(
+    implicit m: Monoid[T],
+    p: Priority[Group[T], Monoid[T]]
+) extends Monoid[Window[T]] {
+  val zero = Window[T](m.zero)
+
+  /*
+   * plus is smart enough to dynamically use a more efficient aggregation
+   * on elements with the Group property.
+   */
+  def plus(a: Window[T], b: Window[T]): Window[T] =
+    p match {
+      case Priority.Preferred(g) => plusG(a, b)(g)
+      case Priority.Fallback(m)  => plusM(a, b)(m)
+    }
+
+  def plusG(a: Window[T], b: Window[T])(implicit g: Group[T]): Window[T] =
+    if (b.items.size >= windowSize) {
+      var total: T = b.total
+      var q = b.items
+      while (q.size > windowSize) {
```

Comment:
Not sure if its possible to do it in this implementation but it would be nice to generalize how elements are evicted from the window. Right now it seems like I can only do fixed windows but I would really like to do time based rolling windows as well. For example: Rolling window of 1 hour and we evict element based on time instead of size

Reply:
My thought was to eventually have two classes/monoids, based on whether we want aggregations over fixed items or fixed time.

```
// Monoid defines aggregation for fixed window length, e.g. last 28 elements. Queue fixed to 28.
case class Window[T](total:T, q: Queue[T])

// Monoid defines aggregation for fixed time, e.g. last 28 days. Queue can grow arbitrarily.
case class WindowWithTs(lastestTime: Double, total: T, q: Queue[(Double, T)])
```

Does that seem sensible?

Reply:
That seems reasonable but I am still not sure why we cant use `ProrityQueue` instead of `Queue`? As a user I would like to control the ordering and if I am passing you a PQ, I can do that by specifying my own comparator